### PR TITLE
Add cutlass support for blackwell fp8 blockwise gemm

### PR DESCRIPTION
**Pull request by wenscarl**
This PR adds support for cutlass blackwell blockwise gemm for fp8.

### CODE REVIEW

**Comment by pathorn** [comment]
It looks like `scaled_mm_blockwise_sm100_fp8.cu` and `scaled_mm_blockwise_sm100_fp8_dispatch.cuh` were committed only as symlinks. When the change is done and ready for review, make sure to double check the actual files are added.

I'm curious to compare it to the blockwise kernels I wrote in the blackwell-rebase-feb20 branch of deepinfra/vllm , since we had to change the scale factor to be fp8

**Comment by kushanam** [comment]
Thank you @wenscarl! Please correct the`cutlass_scaled_mm_supports_block_fp8()` function in `scaled_mm_entry.cu` to allow `>100` arch as well.

**Comment by WoosukKwon** [comment]
@tylertitsworth Can you please take a look?

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8.cu [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                                           torch::Tensor const& a,
+                                           torch::Tensor const& b,
+                                           torch::Tensor const& a_scales,
+                                           torch::Tensor const& b_scales) {
+  int64_t original_rows = a.size(0);
+
+  torch::Tensor a_padded = pad_tensor(a, /*alignment=*/4);
+  torch::Tensor a_scales_padded =
+      pad_tensor(a_scales, /*alignment=*/4, /*col_major=*/true);
+  torch::Tensor out_padded;
```
Comment (tlrmchlsmth) [inline csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8.cu:39]: This dynamic padding could cause subtle performance issues that could be hard to debug. I think it would be best to add a TORCH_CHECK for the alignment and then guarantee alignment by the caller.

We could guarantee alignment here:
https://github.com/vllm-project/vllm/blob/ff38f0a32c5f7a2dd1aa3bee3806d957adacb9bd/vllm/model_executor/layers/quantization/utils/fp8_utils.py#L61-L63

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  static constexpr int ScaleGranularityK =
+      size<2>(MmaTileShape{}) / size<2>(ScalesPerTile{});
+
+  // Shape of the threadblocks in a cluster
+  using ClusterShape_MNK = Shape<_1, _1, _1>;
+
+  // using ScaleConfig =
+  // decltype(cutlass::detail::sm100_trivial_blockwise_scale_config(MmaTileShape_MNK{}));
+  // static constexpr int ScaleGranularityM = size<0>(MmaTileShape{}) /
+  // ScaleMsPerTile;
```
Comment (tlrmchlsmth) [inline csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh:71]: nit: delete commented-out code

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+  using CollectiveMainloop =
+      typename cutlass::gemm::collective::CollectiveBuilder<
+          ArchTag, OperatorClass, ElementA, cute::tuple<LayoutA, LayoutSFA>,
+          AlignmentA, ElementB, cute::tuple<LayoutB, LayoutSFB>, AlignmentB,
+          ElementAccumulator, MmaTileShape, ClusterShape,
+          cutlass::gemm::collective::StageCountAutoCarveout<static_cast<int>(
+              sizeof(typename CollectiveEpilogue::SharedStorage))>,
+          cutlass::gemm::KernelTmaWarpSpecializedBlockwise1SmSm100>::
+          CollectiveOp;
```
Comment (tlrmchlsmth) [inline csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh:102]: light suggestion to turn off clang format for these blocks and manually do the whitespace

**Review by tlrmchlsmth — COMMENTED** [review]
Thanks for the contribution! A couple of comments but looks good overall

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh (hunk 2 above)
Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh:71]: @wenscarl just checking since I see you marked this as resolved: did you forget to push?

Comment (wenscarl) [reply csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh:71]: Thanks for the reminder. I am working on the push, please expect it within next few hours.

File: csrc/cutlass_extensions/common.hpp [hunk 4]
```diff
@@ -59,3 +59,13 @@ struct enable_sm90_only : Kernel {
...
 };
+
+template <typename Kernel>
+struct enable_sm100_or_later : Kernel {
+  template <typename... Args>
+  CUTLASS_DEVICE void operator()(Args&&... args) {
+#if defined __CUDA_ARCH__ && __CUDA_ARCH__ >= 1000
+    Kernel::operator()(std::forward<Args>(args)...);
+#endif
+  }
```
Comment (tlrmchlsmth) [inline csrc/cutlass_extensions/common.hpp:70]: Should we do this instead?
```
template <typename Kernel>
struct enable_sm100_only : Kernel {
  template <typename... Args>
  CUTLASS_DEVICE void operator()(Args&&... args) {
#if defined __CUDA_ARCH__ && __CUDA_ARCH__ == 1000
    Kernel::operator()(std::forward<Args>(args)...);
#endif
  }
```

If I understand correctly, these kernels won't be forward-compatible

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by tlrmchlsmth** [comment]
Looks good to me now, thank you!

Please merge in latest main to fix the `docker-build-image` failure

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+      EpilogueTileShape,
+      ElementAccumulator,
+      ElementCompute,
+      ElementC,
+      LayoutC,
+      AlignmentC,
+      ElementD,
+      LayoutD,
+      AlignmentD,
+      cutlass::epilogue::TmaWarpSpecialized1Sm
```
Comment (thakkarV) [inline csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh:98]: Use 2Sm for compute bound cases

File: csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                                               torch::Tensor const& a_scales,
+                                               torch::Tensor const& b_scales) {
+  auto m = a.size(0);
+  auto k = a.size(1);
+  auto n = b.size(1);
+
+  // Define tile shapes based on the value of m
+  if (m <= 128) {
+    cutlass_gemm_caller_blockwise<cutlass_3x_gemm_fp8_blockwise<
+        OutType, Shape<_64, _128, _128>, Shape<_64, _128, _128>,
```
Comment (thakkarV) [inline csrc/quantization/cutlass_w8a8/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh:192]: Have you tried auto tuning for these?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 7]
```diff
@@ -58,6 +58,15 @@ def apply_w8a8_block_fp8_linear(
...
             shape_supported_by_cutlass = False
     if cutlass_block_fp8_supported and shape_supported_by_cutlass:
+        rows, cols = input_2d.shape
+        should_pad = current_platform.has_device_capability(
+            100) and rows % 4 != 0
+        if should_pad:
+            padding = torch.zeros((4 - (rows % 4), cols),
+                                  dtype=input_2d.dtype,
+                                  device=input_2d.device)
+            input_2d = torch.cat([input_2d, padding], dim=0).contiguous()
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:68]: is this faster than `torch.nn.functional.pad`?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 8]
```diff
@@ -58,6 +58,15 @@ def apply_w8a8_block_fp8_linear(
                 or br not in (1, weight.shape[0])):
             shape_supported_by_cutlass = False
     if cutlass_block_fp8_supported and shape_supported_by_cutlass:
+        rows, cols = input_2d.shape
+        should_pad = current_platform.has_device_capability(
+            100) and rows % 4 != 0
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:63]: can we leave a note about why this is required and when/or-under-which-conditions it can be removed?

**Review by LucasWilkinson — APPROVED** [review]
LGTM now, thanks!
